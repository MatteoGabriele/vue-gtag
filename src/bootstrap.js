import { warn, isFn, loadScript } from "./util";
import { getOptions } from "../src/install";
import optOut from "./api/opt-out";
import pageTracker from "./page-tracker";

export default function() {
  if (typeof document === "undefined" || typeof window === "undefined") {
    return;
  }

  const {
    enabled,
    globalObjectName,
    globalDataLayerName,
    config: { id, params },
    includes,
    pageTrackerEnabled,
    onReady,
    disableScriptLoad
  } = getOptions();

  if (!enabled) {
    optOut();
  }

  if (window[globalObjectName] == null) {
    window[globalDataLayerName] = [];
    window[globalObjectName] = function() {
      window[globalDataLayerName].push(arguments);
    };
  }

  window[globalObjectName]("js", new Date());

  if (params) {
    window[globalObjectName]("config", id, params);
  } else {
    window[globalObjectName]("config", id);
  }

  if (Array.isArray(includes)) {
    includes.forEach(domain => {
      if (domain.params) {
        window[globalObjectName]("config", domain.id, domain.params);
      } else {
        window[globalObjectName]("config", domain.id);
      }
    });
  }

  if (pageTrackerEnabled) {
    pageTracker();
  }

  if (disableScriptLoad) {
    return;
  }

  const domain = "https://www.googletagmanager.com";
  const resource = `${domain}/gtag/js?id=${id}&l=${globalDataLayerName}`;

  return loadScript(resource, domain)
    .then(() => {
      const library = window[globalObjectName];

      if (isFn(onReady)) {
        onReady(library);
      }

      return library;
    })
    .catch(error => {
      warn("Ops! Something happened and gtag.js couldn't be loaded", error);
      return error;
    });
}
