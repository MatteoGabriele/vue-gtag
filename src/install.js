import { noop, mergeDeep } from "./util";
import extend from "./extend";
import bootstrap from "./bootstrap";

let Vue;
let Router;
let options = {
  pageTrackerTemplate: noop,
  onBeforeTrack: noop,
  onAfterTrack: noop,
  onReady: noop,
  enabled: true,
  disableScriptLoad: false,
  bootstrap: true,
  globalObjectName: "gtag",
  pageTrackerEnabled: true,
  pageTrackerScreenviewEnabled: false,
  defaultGroupName: "default",
  includes: null,
  config: null
};

export const getOptions = () => options;
export const setOptions = _options => mergeDeep(options, _options);
export const getVue = () => Vue;
export const getRouter = () => Router;

export function install(_Vue, _options = {}, _Router) {
  Vue = _Vue;
  Router = _Router;

  setOptions(_options);

  extend();

  if (!options.bootstrap) {
    return;
  }

  bootstrap();
}
