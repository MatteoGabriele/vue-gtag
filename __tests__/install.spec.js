import { getVue, getRouter, install, getOptions, setRouter } from "@/install";
import bootstrap from "@/bootstrap";
import extend from "@/extend";

jest.mock("@/extend");
jest.mock("@/bootstrap");

describe("install", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should have default options", () => {
    expect(getOptions()).toMatchSnapshot();
  });

  it("should return VueRouter", () => {
    install(null, null, { foo: "bar" });
    expect(getRouter()).toEqual({ foo: "bar" });
  });

  it("should set VueRouter", () => {
    const router = { foo: "bar" };

    install();
    setRouter(router);

    expect(getRouter()).toEqual(router);
  });

  it("should return Vue", () => {
    install({ foo: "bar" });
    expect(getVue()).toEqual({ foo: "bar" });
  });

  it("should call extend", () => {
    install();
    expect(extend).toHaveBeenCalled();
  });

  it("should call bootstrap", () => {
    install();
    expect(bootstrap).toHaveBeenCalled();
  });

  it("should not bootstrap", () => {
    install(null, { bootstrap: false });
    expect(bootstrap).not.toHaveBeenCalled();
  });
});
