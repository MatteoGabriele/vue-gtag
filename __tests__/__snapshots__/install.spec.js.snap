// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`install should have default options 1`] = `
Object {
  "bootstrap": true,
  "config": null,
  "defaultGroupName": "default",
  "disableScriptLoad": false,
  "enabled": true,
  "globalObjectName": "gtag",
  "includes": null,
  "onAfterTrack": [Function],
  "onBeforeTrack": [Function],
  "onReady": [Function],
  "pageTrackerEnabled": true,
  "pageTrackerScreenviewEnabled": false,
  "pageTrackerTemplate": [Function],
}
`;
